### PR TITLE
Added support for 'Tensor.clipped(min:max:)' and its VJP.

### PR DESCRIPTION
**Pull request by eaplatanios**

### CODE REVIEW

File: Sources/TensorFlow/Operators/Math.swift [hunk 1]
```diff
@@ -494,6 +494,58 @@ public extension Tensor where Scalar == Bool {
     }
 }
 
+public extension Tensor where Scalar: TensorFlowNumeric {
+    /// Returns `max(min(self, max), min)`.
+    @inlinable
+    @differentiable(vjp: _vjpClipping where Scalar: TensorFlowFloatingPoint)
+    func clipping(min: Tensor, max: Tensor) -> Tensor {
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]: “clipping“ -> “clipped”

**Comment by Shashi456** [comment]
I guess after this PR, a lot of layers can be updated to avoid certain values beyond bounds. Like its done in keras.

File: Sources/TensorFlow/Operators/Math.swift (hunk 1 above)
Comment (eaplatanios) [reply Sources/TensorFlow/Operators/Math.swift]: Done.

**Comment by eaplatanios** [comment]
> I guess after this PR, a lot of layers can be updated to avoid certain values beyond bounds. Like its done in keras.

Could you give examples of some of these layers in Keras? I'm curious to see where they clip implicitly.

**Comment by rxwei** [comment]
I feel the tensor-scalar and scalar-tensor overloads are not quite necessary and may not set a good example to follow for future APIs. Say, if some API has three 0-D tensor parameters (e.g. batch norm), would we want to write all 6 combinations of scalars parameters and tensor parameters?

**Comment by eaplatanios** [comment]
I agree. Would it be possible to make `Scalar: TensorFlowScalar` implicitly convertible to `Tensor<Scalar>`?

**Comment by rxwei** [comment]
No, implicit conversion is generally not possible in Swift except for literals. However, scalar literal conversion for `Tensor` has been very problematic.

**Comment by eaplatanios** [comment]
What is it that makes it problematic? Is it too inefficient for the type checker?

**Comment by rxwei** [comment]
Type checking ambiguity. `s + t` is possible where `s` is a scalar and `t` is a tensor, but if you make tensor also expressible by literal, the type checker won't know which overload you are calling for an expression like `3 + x` (scalar-tensor or tensor-tensor).

**Comment by eaplatanios** [comment]
Can we not impose a hierarchy on the literal conversion rules?

**Comment by rxwei** [comment]
What do you mean exactly?

**Comment by rxwei** [comment]
Type checking precedence changes generally need to be an evolution proposal.

**Comment by eaplatanios** [comment]
Yes that would need an evolution proposal indeed. We could go for something even more powerful and support implicit type conversions using something like this:
```swift
@implicitConversion(precedence: ...)
func _<Scalar: TensorFlowScalar>(_ value: Scalar) -> Tensor<Scalar> {
  ...
}
```
Do you think something like this could potentially go through evolution? Also, how time consuming would it be support something like this in the compiler?

**Comment by eaplatanios** [comment]
Note that this would be separate from how literals are handled.

**Comment by rxwei** [comment]
This would be highly unlikely. I would suggest asking a question on the Swift evolution forum about it.

**Comment by eaplatanios** [comment]
Started a discussion thread [here](https://forums.swift.org/t/implicit-conversions/27026).
